lib/sqlgen.js

```diff
@@ -0,0 +1,249 @@
+    return this;
+  }
+
+  whereNot(key, cond, value) {
+    this.operations.get('where').push(this._where(key, cond, value, 'NOT'));
+    return this;
+  }
+
+  whereNull(key) {
+    this.operations.get('where').push({ key, cond: 'IS', value: 'null' });
+    return this;
+  }
+
+  whereNotNull(key) {
+    this.operations.get('where').push({
+      key,
+      cond: 'IS',
+      value: 'null',
+      mod: 'NOT',
+    });
+    return this;
+  }
+
+  _whereIn(key, conds, mod) {
+    const value = `(${conds.map(escapeValue).join(', ')})`;
+    return { key, cond: 'IN', value, mod };
+  }
+
+  whereIn(key, conds) {
+    this.operations.get('where').push(this._whereIn(key, conds));
+    return this;
+  }
+
+  whereNotIn(key, conds) {
+    this.operations.get('where').push(this._whereIn(key, conds, 'NOT'));
+    return this;
+  }
+
+  orderBy(field, dir = 'ASC') {
+    dir = dir.toUpperCase();
+    this.operations.get('orderBy').add({ field, dir });
+    return this;
+  }
+
+  groupBy(...fields) {
+    const groupBy = this.operations.get('groupBy');
+    iter(fields).forEach(f => groupBy.add(f));
+    return this;
+  }
+
+  limit(limit) {
+    checkType(limit, 'limit', 'number');
+    this.operations.set('limit', limit);
+    return this;
+  }
+
+  offset(offset) {
+    checkType(offset, 'offset', 'number');
+    this.operations.set('offset', offset);
+    return this;
+  }
+
+  count(field = '*') {
+    this.operations.get('count').push({ field });
+    return this;
+  }
+
+  avg(field) {
+    this.operations.get('avg').push({ field });
+    return this;
+  }
+
+  min(field) {
+    this.operations.get('min').push({ field });
+    return this;
+  }
+
+  max(field) {
+    this.operations.get('max').push({ field });
+    return this;
+  }
+
+  sum(field) {
+    this.operations.get('sum').push({ field });
+    return this;
+  }
+
+  toString() {
+    let query = 'SELECT';
+
+    if (this.operations.get('selectDistinct')) query += ' DISTINCT';
+
+    // handle initial select clause
+    const select = this.operations.get('select');
+    if (select.size > 0) {
+      query += ' ' + iter(select).reduce((acc, id) => acc + ', ' + id);
+    } else {
+      query += ' *';
+    }
+
+    // handle operations
+    for (const fn of Object.keys(functionHandlers)) {
+      const ops = this.operations.get(fn);
+      if (ops.length === 0) continue;
+      else if (query.endsWith(' *')) query = query.slice(0, -2);
+      else query += ',';
+      const handler = functionHandlers[fn];
+      // eslint-disable-next-line no-loop-func
+      query += ops.reduce((acc, op) => acc + handler(op, query) + ',', ' ')
+        .slice(0, -1);
+    }
+
+    // handle from
+    const tableName = this.operations.get('from');
+    if (!tableName) {
+      throw new Error('Cannot generate SQL, tableName is undefined');
+    }
+    query += ` FROM ${tableName}`;
+
+    // handle where clauses
+    const whereClauses = this.operations.get('where');
```

Comment:
Decompose all clauses to separate functions, `const templates = { where, group, order... }`